challenge_8/scala/ndsmith3/WeirdLinkedList.scala

```diff
@@ -0,0 +1,87 @@
+import scala.util.Random.nextInt
+
+case class WeirdList[A](data: A) {
+  var next: Option[WeirdList[A]] = None
+  var rand: Option[WeirdList[A]] = None
+  var length: Int                = 1
+
+  def add(data: A): Unit = {
+    def assignRand(list: WeirdList[A])(
+        index: Int = nextInt(this.length)): Unit = {
+      if (index == this.indexOf(list.data)) assignRand(list)()
+      else list.rand = Option(this.subList(index))
+    }
+
+    if (this.next.isDefined) this.next.get.add(data)
+    else this.next = Option(WeirdList(data))
+    if (this.length > 1) this foreach (x => assignRand(x)())
+    length += 1
+  }
+
+  def addNode(node: WeirdList[A]): Unit = {
+    if (this.next.isDefined) this.next.get.addNode(node)
+    else this.next = Option(node)
+  }
+
+  def indexOf(data: A, index: Int = 0): Int = {
+    if (this.data == data) index
+    else this.next.get.indexOf(data, index + 1)
+  }
+
+  def subList(index: Int): WeirdList[A] =
+    (1 to index).foldLeft(this)((x, _) => x.next.get)
+
+  def foreach(f: WeirdList[A] => Unit): Unit = {
+    f(this)
+    if (this.next.isDefined) this.next.get.foreach(f)
```

Comment:
> if (this.next.isDefined) this.next.get.foreach(f)

Ouch!  Instead of checking if it's defined and then calling get, you should just use this.next.foreach(_.foreach(f)) which does nothing if this.next is None.  It's a little unfortunate that the 2 foreach's have the same name, but oh well.